### DISCUSSION TITLE
Island Detection

### DISCUSSION THREAD

**Discussion by FNGgames** [libvips/libvips#3523] [Q&A] [answered]
Hi, 

Can anyone direct me towards a libvips approach to island detection? The result im looking for is an image where its contiguous non-zero patches are individually colored so they can be easily filtered and seperated island by island for further processing.

For example:

![IslandDetector-Before](https://github.com/libvips/libvips/assets/18085014/4a744446-2bcd-4768-a57c-15763b3f4b47)

![IslandDetector-After](https://github.com/libvips/libvips/assets/18085014/9b388eed-e93e-4503-a3d4-d266e1937d89)

Ideally I'd love to also get out some data about how many islands were detected, the colors used to label them etc.

**Comment by jcupitt** [comment, marked as answer]
Hi @FNGgames,

It has a couple of things for this.

`labelregions` scans an image and gives 8-way connected pixels of the same value a unique number. Your sample PNG is antialiased, so it doesn't label well. If you threshold it first you get a nice result:

```
$ vips relational_const sample.png t1.v more 128
$ vips labelregions t1.v t2.v
$ vips scale t2.v t3.v
$ vips falsecolour t3.v x.png
```

To make:

![x](https://github.com/libvips/libvips/assets/580843/c5f8d809-e035-42f2-9e30-21f3ecc3ff26)

`labelregions` will give regions labels like 1, 2, 3 etc., so they are very dark. If you scale and falsecolour they become easy to see. It's a bit less painful in eg. python:

```python
result = (image > 128).labelregions().scale().falsecolour()
```

**Reply by FNGgames** [reply]
This is perfect, thankyou.

**Comment by jcupitt** [comment]
You can find stats on labels with `hist_find_indexed`.  This will compute a histogram, but use an index image rather than an image of values.

For example, you could make an image where each pixel is its own coordinate, then calling `hist_find_indexed` on that plus the output of `labelregions` would sum the (x, y) positions of each label. If you divide by the regular histogram of the label image, you get the centre of mass of each region in an array.

You can compute eg. bounding boxes in the same way.

**Comment by jcupitt** [comment]
The operation behind `labelregions` is `draw_flood`, a fast scanline-based flood fill. You can call that yourself to implement other operations on connected regions.

**Reply by FNGgames** [reply]
Awesome, thanks very much @jcupitt
